### PR TITLE
mgr: dashboard GUI module

### PR DESCRIPTION
**Pull request by dmick**
Cleanup of work from @jcsp to add a ceph-mgr Python module to provide a web service that displays a simple GUI status, with one feature added to allow configuration of the listen addr and port for the server.

Configuration is done with ceph config-key put; the key names are 

mgr.dashboard.server_addr
mgr.dashboard.server_port

### CODE REVIEW

**Comment by liewegas** [comment]
Since you can have multiple mgr's on different hosts, I think we need to qualify these with the mgr id.  Something like

mgr/x/dashboard/server_addr
mgr/x/dashboard/server_port
mgr/y/dashboard/server_addr
mgr/y/dashboard/server_port

(The dm-crypt stuff in config-key is currently using / instead of . as delimiter; we may as well stick with that?)

**Comment by dmick** [comment]
ok, will hack

**Comment by dmick** [comment]
Working on Yet Another Rebase

**Comment by dmick** [comment]
Removing some commits that had made it into master in the meantime made this merge more cleanly; this branch appears to pass dumb functional tests again.

**Comment by dmick** [comment]
uhh....there are apparently a few more files than I wanted here.

**Comment by jcsp** [comment]
@dmick you could probably cull a lot of the AdminLTE guff, I think there are various optional bits (themes/plugins) in there where we only need the files for the ones we use.

**Comment by dmick** [comment]
rebased again

**Comment by liewegas** [comment]
weird thing on this run: http://pulpito.ceph.com/sage-2017-05-17_15:24:19-rados-wip-sage-testing---basic-smithi/1189010

smithi055 and smithi067 are both trusty, but the cherrypy error only came up on one of them.  not a package dependency for deb yet?

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-17_15:24:19-rados-wip-sage-testing---basic-smithi/
looks fine in rados suite, fwiw.

**Comment by tserong** [comment]
specfile change LGTM FWIW

**Comment by liewegas** [comment]
Ok, I suggest we merge this as is and refine on top of master.

File: src/mgr/PyState.cc [hunk 1]
```diff
@@ -196,10 +196,10 @@ ceph_config_get(PyObject *self, PyObject *args)
   std::string value;
   bool found = global_handle->get_config(handle, what, &value);
   if (found) {
-    derr << "Found" << dendl;
+    derr << "ceph_config_get " << what << " found: " << value.c_str() << dendl;
```
Comment (tchaikov) [inline src/mgr/PyState.cc]: could you use `__func__` instead here?

File: src/mgr/PyState.cc [hunk 2]
```diff
@@ -196,10 +196,10 @@ ceph_config_get(PyObject *self, PyObject *args)
...
     return PyString_FromString(value.c_str());
   } else {
-    derr << "Not found" << dendl;
+    derr << "ceph_config_get " << what << " not found " << dendl;
```
Comment (tchaikov) [inline src/mgr/PyState.cc]: ditto.

**Comment by tchaikov** [comment]
@dmick please feel free to merge this when you feel appropriate.

**Comment by dmick** [comment]
Dependencies added, very simple smoke test added to make check

**Comment by liewegas** [comment]
<pre>/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mgr/mgr-dashboard-smoke.sh:45: run:  sleep 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mgr/mgr-dashboard-smoke.sh:39: run:  [[ 4 < 30 ]]
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mgr/mgr-dashboard-smoke.sh:47: run:  curl -s http://127.0.0.1:7001/toplevel_data
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mgr/mgr-dashboard-smoke.sh:48: run:  jq .health.overall_status
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mgr/mgr-dashboard-smoke.sh:48: run:  grep HEALTH_
parse error: Invalid numeric literal at line 1, column 10
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mgr/mgr-dashboard-smoke.sh:48: run:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1692: main:  display_logs td/mgr-dashboard-smoke
</pre>

**Comment by dmick** [comment]
Yep; internal failure in dashboard.  Trying to reproduce now.

**Comment by dmick** [comment]
It was intermittent; sometimes after the mgr said 'available', dashboard still wasn't quite ready.  Added a second retry loop; running with `ctest --output-on-failure --repeat-until-fail 100 -R mgr-dash` looks solid now

**Comment by liewegas** [comment]
excellent, thanks!

**Comment by dmick** [comment]
Sorry @tchaikov, missed your comments about __func__; I wasn't ignoring you on purpose

**Comment by tchaikov** [comment]
@dmick no worries, we can always improve it once we are at it.

**Pull request merged**